apps/voting/app/src/vote-hooks.js

```diff
@@ -0,0 +1,80 @@
+import { useState, useEffect } from 'react'
+import { useAragonApi } from '@aragon/api-react'
+
+// when `params` are updated, call `fn` and pass the result
+function usePromiseResult(fn, params, memoParams, defaultValue) {
+  const [result, setResult] = useState(defaultValue)
+  useEffect(() => {
+    let cancelled = false
+    fn(...params).then(value => {
+      if (!cancelled) {
+        setResult(value)
+      }
+    })
+    return () => {
+      cancelled = true
+    }
+  }, memoParams)
+  return result
+}
+
+async function getUserBalance(vote, userAccount, tokenContract, tokenDecimals) {
+  if (!tokenContract || !userAccount) {
+    return -1
+  }
+
+  const balance = await tokenContract
+    .balanceOfAt(userAccount, vote.data.snapshotBlock)
+    .toPromise()
+
+  return Math.floor(parseInt(balance, 10) / Math.pow(10, tokenDecimals))
+}
+
+async function getCanVote(vote, userAccount, api) {
+  if (!vote) {
+    return false
+  }
+
+  // If the account is not present, we assume the account is not connected.
+  if (!userAccount) {
+    return vote.data.open
+  }
+
+  return api.call('canVote', vote.voteId, userAccount).toPromise()
+}
+
+async function getCanExecute(vote, api) {
+  if (!vote) {
+    return false
+  }
+  return api.call('canExecute', vote.voteId).toPromise()
+}
+
+export const useCurrentVoteData = (
+  vote,
+  userAccount,
+  tokenContract,
+  tokenDecimals
+) => {
+  const { api } = useAragonApi()
+  return {
```

Comment:
Perhaps we could bail here with the defaults if `!vote`, to simplify the logic afterwards?

Reply:
[Forbidden :no_good_man: :no_entry: :no_entry_sign: ](https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level)

Doing so would prevent React to keep track of the hooks and their corresponding components.